### PR TITLE
WT-5448 Fix column reconciliation overwrites cells with default time pairs

### PR DESCRIPTION
**Pull request by quchenhao**
The code is complex and I don't understand all of it. So please have a careful look.

### CODE REVIEW

File: src/reconcile/rec_col.c [hunk 1]
```diff
@@ -760,17 +760,7 @@ __wt_rec_col_var(
             if (ins != NULL && WT_INSERT_RECNO(ins) == src_recno) {
                 WT_ERR(__wt_rec_upd_select(session, r, ins, cip, vpack, &upd_select));
                 upd = upd_select.upd;
-                if (upd == NULL) {
```
Comment (quchenhao) [inline src/reconcile/rec_col.c]: Here upd equals null means we are writing the same value to disk thus the time pairs should be unchanged.

File: src/reconcile/rec_col.c [hunk 2]
```diff
@@ -963,11 +950,8 @@ __wt_rec_col_var(
             upd = upd_select.upd;
             n = WT_INSERT_RECNO(ins);
         }
+        /* Deletion of uncommitted keys are visible to all. */
```
Comment (quchenhao) [inline src/reconcile/rec_col.c]: When walking the append list, if we don't select an update to write to disk, there must be only uncommitted updates on that key. In that case, we write a deleted key to disk and the deletion should be globally visible.

File: src/reconcile/rec_col.c [hunk 3]
```diff
@@ -1002,19 +986,9 @@ __wt_rec_col_var(
                     rle += skip;
                     src_recno += skip;
                 }
-            } else if (upd == NULL) {
```
Comment (quchenhao) [inline src/reconcile/rec_col.c]: We already set the time pairs so no need to set them again here.

File: src/reconcile/rec_row.c [hunk 4]
```diff
@@ -839,34 +837,6 @@ __wt_rec_row_leaf(
                   stop_txn, 0));
                 dictionary = true;
                 break;
-            case WT_UPDATE_TOMBSTONE:
```
Comment (quchenhao) [inline src/reconcile/rec_row.c]: We never select a tombstone to write to disk in durable history.

**Comment by quchenhao** [comment]
retest this please

File: src/reconcile/rec_col.c [hunk 5]
```diff
@@ -963,11 +950,8 @@ __wt_rec_col_var(
             upd = upd_select.upd;
             n = WT_INSERT_RECNO(ins);
         }
+        /* Deletion of uncommitted keys are visible to all. */
         if (upd == NULL) {
-            /*
-             * TIMESTAMP-FIXME I'm pretty sure this is wrong: a NULL update means an item was
-             * deleted, and I think that requires a tombstone on the page.
-             */
```
[the remark this answers is not in the document]
Comment (quchenhao) [reply src/reconcile/rec_col.c]: That means the updates to that key are not committed.

File: src/reconcile/rec_col.c [hunk 6]
```diff
@@ -1002,25 +986,9 @@ __wt_rec_col_var(
...
-
+            } else if (upd == NULL)
                 deleted = true;
-            } else {
-                durable_ts = upd_select.durable_ts;
-                start_ts = upd_select.start_ts;
-                start_txn = upd_select.start_txn;
-                stop_ts = upd_select.stop_ts;
-                stop_txn = upd_select.stop_txn;
-
```
[the remark this answers is not in the document]
Comment (quchenhao) [reply src/reconcile/rec_col.c]: fixed

**Comment by quchenhao** [comment]
I don't know how to write a test for this as I can't check time pairs in cell from python. I put some asserts in the code instead.

**Comment by quchenhao** [comment]
This is ready to be reviewed again.

File: src/reconcile/rec_col.c [hunk 7]
```diff
@@ -760,17 +760,11 @@ __wt_rec_col_var(
...
-                    start_ts = WT_TS_NONE;
-                    start_txn = WT_TXN_NONE;
-                    stop_ts = WT_TS_MAX;
-                    stop_txn = WT_TXN_MAX;
-                } else {
+                /*
+                 * If upd == NULL, we are writing the same key to disk. The time pairs should come
+                 * from the cell itself.
+                 */
+                if (upd != NULL) {
```
Comment (keithbostic) [inline src/reconcile/rec_col.c]: I don't understand the phrase `the same key`. What is the key the same as?

File: src/reconcile/rec_col.c [hunk 8]
```diff
@@ -882,6 +873,15 @@ __wt_rec_col_var(
...
                         memcmp(last.value->data, data, size) == 0))) {
+                    /*
+                     * The start time pair for deleted keys must be (WT_TS_NONE, WT_TXN_NONE) and
+                     * stop time pair must be (WT_TS_MAX, WT_TXN_MAX) since we no longer select
+                     * tombstone to write to disk and the deletion of the keys must be globally
+                     * visible.
+                     */
+                    WT_ASSERT(session, (!deleted && !last.deleted) ||
+                        (last.start_ts == WT_TS_NONE && last.start_txn == WT_TXN_NONE &&
+                                         last.stop_ts == WT_TS_MAX && last.stop_txn == WT_TXN_MAX));
```
Comment (keithbostic) [inline src/reconcile/rec_col.c]: Is this true, even when we took a deleted record from an update? So, the user does `WT_CURSOR.remove`, that puts an update on the chain that is a "deleted" record. Will `last.XXX` be `NONE/MAX`?

Comment (quchenhao) [reply src/reconcile/rec_col.c]: That currently is reconciled as a normal record on the page with a stop timestamp. We haven't decided how to remove them from the disk image yet.

Comment (quchenhao) [reply src/reconcile/rec_col.c]: If we can remove the record from the disk image, then the deletion must be globally visible. We can set its start time pair to (0, 0) and stop time pair to (MAX, MAX).

File: src/reconcile/rec_col.c (hunk 1 above)
Comment (quchenhao) [reply src/reconcile/rec_col.c]: My understanding here is wrong. This is in the insert list so we are also handling new keys here.

**Comment by quchenhao** [comment]
@keithbostic I realized that I misunderstood some part of the code. Please have a look at this again.

File: src/reconcile/rec_col.c [hunk 9]
```diff
@@ -817,8 +798,23 @@ __wt_rec_col_var(
...
+                    start_ts = WT_TS_NONE;
+                    start_txn = WT_TXN_NONE;
+                    stop_ts = WT_TS_MAX;
+                    stop_txn = WT_TXN_MAX;
+
                     goto compare;
+                }
+
+                /* The key on the old disk image is unchanged. Use time pairs from the cell. */
+                durable_ts = newest_durable_ts;
```
[the remark this answers is not in the document]
Comment (quchenhao) [reply src/reconcile/rec_col.c]: Not sure. Will check.

[the remark this answers is not in the document]
Comment (bvpvamsikrishna) [reply src/reconcile/rec_col.c]: We don't have durable timestamp in cell. So we cannot use. This is the problem we need to solve in prepare support for durable history.
Could you update comments to indicate to check this as part of PM-1524.

[the remark this answers is not in the document]
Comment (quchenhao) [reply src/reconcile/rec_col.c]: For keys not changed, we just reuse the previous newest durable ts onpage. Later the newest_durable_ts will be updated downstream.

**Comment by quchenhao** [comment]
> lgtm.
> with suggestion to update the comment.

Sure. We have newest_durable_ts on the page but not sure if that is helpful though.

**Pull request merged**